File: packages/integrations/anthropic-js/src/session.ts
import { randomUUID } from "crypto";
import type { AgentEndEvent } from "@agentmetrics/core";
import { estimateCost as sharedEstimateCost } from "@agentmetrics/core";


const MAX_RETRY_ATTEMPTS = 3;

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function sendWithRetry(
  url: string,
  apiKey: string,
  payload: object,
): Promise<void> {
  const body = JSON.stringify(payload);
  for (let attempt = 0; attempt < MAX_RETRY_ATTEMPTS; attempt++) {
    try {
      const res = await fetch(`${url}/v1/events`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${apiKey}`,
        },
        body,
      });
      if (res.ok || (res.status >= 400 && res.status < 500)) return;
      if (attempt < MAX_RETRY_ATTEMPTS - 1) {
        await sleep(Math.min(1000 * 2 ** attempt + Math.random() * 200, 10_000));
      }
    } catch {
      if (attempt < MAX_RETRY_ATTEMPTS - 1) {
        await sleep(Math.min(1000 * 2 ** attempt + Math.random() * 200, 10_000));
      }
    }
  }
}


interface SessionState {
  agentId:          string;
  sessionId:        string;
  startMs:          number;
  inputTokens:      number;
  outputTokens:     number;
  cacheReadTokens:  number;
  cacheWriteTokens: number;
  llmCalls:         number;
  toolCalls:        number;
  toolErrors:       number;
  toolNames:        Set<string>;
  model:            string | null;
  status:           "success" | "failed";
  error:            string | null;
}

function newSessionState(agentId: string, sessionId: string): SessionState {
  return {
    agentId,
    sessionId,
    startMs:          performance.now(),
    inputTokens:      0,
    outputTokens:     0,
    cacheReadTokens:  0,
    cacheWriteTokens: 0,
    llmCalls:         0,
    toolCalls:        0,
    toolErrors:       0,
    toolNames:        new Set(),
    model:            null,
    status:           "success",
    error:            null,
  };
}

function processEvent(state: SessionState, event: Record<string, unknown>): boolean {
  const etype = (event["type"] as string | undefined) ?? "";

  if (etype === "span.model_request_end") {
    const usage = (event["usage"] as Record<string, number> | null) ?? {};
    state.inputTokens       += usage["input_tokens"]              ?? 0;
    state.outputTokens      += usage["output_tokens"]             ?? 0;
    state.cacheReadTokens   += usage["cache_read_input_tokens"]   ?? 0;
    state.cacheWriteTokens  += usage["cache_creation_input_tokens"] ?? 0;
    if (!state.model && event["model"]) {
      state.model = event["model"] as string;
    }
    state.llmCalls++;

  } else if (
    etype === "agent.tool_use" ||
    etype === "agent.mcp_tool_use" ||
    etype === "agent.custom_tool_use"
  ) {
    state.toolCalls++;
    const name = (event["tool_name"] ?? event["name"]) as string | undefined;
    if (name) state.toolNames.add(name);

  } else if (etype === "agent.tool_result") {
    const content = event["content"] as Record<string, unknown> | null;
    if (content?.["is_error"]) state.toolErrors++;

  } else if (etype === "session.error") {
    state.status = "failed";
    state.error  = String(event["error"] ?? "").slice(0, 500) || null;

  } else if (etype === "session.status_terminated") {
    return true;
  }

  return false;
}

async function emit(
  baseUrl: string,
  apiKey: string,
  state: SessionState,
): Promise<void> {
  const durationMs = performance.now() - state.startMs;
  const est = sharedEstimateCost(
    state.model ?? undefined,
    state.inputTokens,
    state.outputTokens,
    state.cacheReadTokens,
    state.cacheWriteTokens,
  );

  const payload: AgentEndEvent = {
    event_id:                 randomUUID(),
    trace_id:                 state.sessionId,
    agent_id:                 state.agentId,
    platform:                 "anthropic",
    event_name:               "agent_end",
    ts:                       Date.now(),
    redaction_policy_version: "v1-strict",
    status:      state.status,
    duration_ms: Math.round(durationMs * 100) / 100,
    tool_calls:  state.toolCalls,
    tool_errors: state.toolErrors,
    tool_names:  [...state.toolNames],
    llm_calls:   state.llmCalls,
    input_tokens:  state.inputTokens,
    output_tokens: state.outputTokens,
    ...(state.model            ? { model:              state.model            } : {}),
    ...(state.error            ? { error:              state.error            } : {}),
    ...(state.cacheReadTokens  ? { cache_read_tokens:  state.cacheReadTokens  } : {}),
    ...(state.cacheWriteTokens ? { cache_write_tokens: state.cacheWriteTokens } : {}),
    ...(est != null            ? { estimated_cost_usd: est                    } : {}),
  };

  sendWithRetry(baseUrl, apiKey, payload).catch(() => {});
}


/**
 * Wraps a Claude Managed Agents session stream and tracks observability data.
 *
 * Usage (async iterator)::
 *
 *   import { AgentMetricsSessionTracker } from "agentmetrics-anthropic";
 *
 *   const tracker = new AgentMetricsSessionTracker({ apiKey: "am_..." });
 *
 *   await tracker.track(client, sessionId, async (stream) => {
 *     for await (const event of stream) {
 *       // handle events as normal
 *     }
 *   });
 *
 * Or wrapping an existing stream::
 *
 *   const rawStream = client.beta.sessions.events.stream(sessionId);
 *   const tracked   = tracker.wrap(rawStream, sessionId);
 *   for await (const event of tracked) { ... }
 */
export class AgentMetricsSessionTracker {
  private readonly _apiKey:  string;
  private readonly _agentId: string;
  private readonly _baseUrl: string;

  constructor(opts: {
    apiKey:   string;
    agentId?: string;
    baseUrl?: string;
  }) {
    this._apiKey  = opts.apiKey;
    this._agentId = opts.agentId ?? "anthropic-agent";
    this._baseUrl = opts.baseUrl ?? "http://localhost:8099";
  }

  /**
   * Wraps any async iterable of session events with tracking.
   * Emits metrics when `session.status_terminated` is received or the
   * iterator exhausts.
   */
  wrap(
    rawStream: AsyncIterable<Record<string, unknown>>,
    sessionId: string,
  ): AsyncIterable<Record<string, unknown>> {
    const state   = newSessionState(this._agentId, sessionId);
    const baseUrl = this._baseUrl;
    const apiKey  = this._apiKey;

    return {
      [Symbol.asyncIterator](): AsyncIterator<Record<string, unknown>> {
        const iter = rawStream[Symbol.asyncIterator]();
        let done = false;
        return {
          async next() {
            if (done) return { value: undefined, done: true };
            const result = await iter.next();
            if (result.done) {
              if (!done) {
                done = true;
                await emit(baseUrl, apiKey, state);
              }
              return { value: undefined, done: true };
            }
            const event = result.value;
            const terminated = processEvent(state, event);
            if (terminated) {
              done = true;
              await emit(baseUrl, apiKey, state);
            }
            return { value: event, done: false };
          },
          async return(value) {
            if (!done) {
              done = true;
              state.status = "failed";
              state.error  = "stream cancelled";
              await emit(baseUrl, apiKey, state);
            }
            return iter.return ? iter.return(value) : { value, done: true };
          },
        };
      },
    };
  }

  /**
   * Higher-level helper: opens a session event stream, wraps it with
   * tracking, and passes the tracked stream to your callback.
   *
   * @param client     - Anthropic SDK client (with beta.sessions.events.stream)
   * @param sessionId  - The session ID to stream events for
   * @param fn         - Async callback receiving the tracked stream
   */
  async track<T>(
    client: {
      beta: {
        sessions: {
          events: {
            stream(sessionId: string, ...args: unknown[]): AsyncIterable<Record<string, unknown>>;
          };
        };
      };
    },
    sessionId: string,
    fn: (stream: AsyncIterable<Record<string, unknown>>) => Promise<T>,
    ...streamArgs: unknown[]
  ): Promise<T> {
    const state   = newSessionState(this._agentId, sessionId);
    const baseUrl = this._baseUrl;
    const apiKey  = this._apiKey;
    let emitted   = false;

    const rawStream = client.beta.sessions.events.stream(sessionId, ...streamArgs);

    const trackedStream: AsyncIterable<Record<string, unknown>> = {
      [Symbol.asyncIterator](): AsyncIterator<Record<string, unknown>> {
        const iter = rawStream[Symbol.asyncIterator]();
        let done = false;
        return {
          async next() {
            if (done) return { value: undefined, done: true };
            const result = await iter.next();
            if (result.done) {
              done = true;
              return { value: undefined, done: true };
            }
            const event = result.value;
            const terminated = processEvent(state, event);
            if (terminated) done = true;
            return { value: event, done: false };
          },
        };
      },
    };

    try {
      const result = await fn(trackedStream);
      if (!emitted) {
        emitted = true;
        await emit(baseUrl, apiKey, state);
      }
      return result;
    } catch (err) {
      if (!emitted) {
        emitted = true;
        state.status = "failed";
        state.error  = String(err).slice(0, 500);
        await emit(baseUrl, apiKey, state);
      }
      throw err;
    }
  }
}

File: packages/integrations/anthropic/agentmetrics_anthropic/session.py
from __future__ import annotations

import time
from typing import Any

from agentmetrics.http_client import HttpClient
from agentmetrics_shared import AgentEndEvent, estimate_cost


class _SessionState:
    __slots__ = (
        "_emitted",
        "agent_id",
        "cache_read_tokens",
        "cache_write_tokens",
        "error",
        "input_tokens",
        "llm_calls",
        "model",
        "output_tokens",
        "session_id",
        "start_ms",
        "status",
        "tool_calls",
        "tool_errors",
        "tool_names",
    )

    def __init__(self, agent_id: str, session_id: str) -> None:
        self.agent_id   = agent_id
        self.session_id = session_id
        self.start_ms   = time.monotonic()
        self.input_tokens     = 0
        self.output_tokens    = 0
        self.cache_read_tokens  = 0
        self.cache_write_tokens = 0
        self.llm_calls   = 0
        self.tool_calls  = 0
        self.tool_errors = 0
        self.tool_names: set[str] = set()
        self.model: str | None = None
        self.status = "success"
        self.error: str | None = None
        self._emitted = False

    def absorb_span_end(self, event: Any) -> None:
        """Extract token counts from span.model_request_end event."""
        usage = getattr(event, "usage", None) or {}
        if isinstance(usage, dict):
            self.input_tokens       += usage.get("input_tokens", 0) or 0
            self.output_tokens      += usage.get("output_tokens", 0) or 0
            self.cache_read_tokens  += usage.get("cache_read_input_tokens", 0) or 0
            self.cache_write_tokens += usage.get("cache_creation_input_tokens", 0) or 0
        else:
            self.input_tokens       += getattr(usage, "input_tokens", 0) or 0
            self.output_tokens      += getattr(usage, "output_tokens", 0) or 0
            self.cache_read_tokens  += getattr(usage, "cache_read_input_tokens", 0) or 0
            self.cache_write_tokens += getattr(usage, "cache_creation_input_tokens", 0) or 0
        if not self.model:
            self.model = getattr(event, "model", None)
        self.llm_calls += 1

    def absorb_tool(self, event: Any, error: bool = False) -> None:
        self.tool_calls += 1
        if error:
            self.tool_errors += 1
        name = getattr(event, "tool_name", None) or getattr(event, "name", None)
        if name:
            self.tool_names.add(str(name))


def _process_event(state: _SessionState, event: Any) -> bool:
    """
    Process one SSE event from the Managed Agents session stream.
    Returns True if the session has terminated and the run is complete.
    """
    etype = getattr(event, "type", "") or ""

    # LLM call completed - extract tokens
    if etype == "span.model_request_end":
        state.absorb_span_end(event)

    # Tool call
    elif etype in ("agent.tool_use", "agent.mcp_tool_use", "agent.custom_tool_use"):
        state.absorb_tool(event)

    # Tool result error
    elif etype == "agent.tool_result":
        content = getattr(event, "content", None)
        is_error = getattr(content, "is_error", False) if content else False
        if is_error:
            state.tool_errors += 1

    # Session error
    elif etype == "session.error":
        state.status = "failed"
        state.error  = str(getattr(event, "error", ""))[:500] or None

    # Session terminated → run complete
    elif etype == "session.status_terminated":
        state.status = "failed"
        # INT-13: if terminated due to error (not normal completion), count it
        if getattr(event, "error", None) is not None:
            state.tool_errors = getattr(state, "tool_errors", 0) + 1
        return True

    return False


class AgentMetricsSessionTracker:
    """
    Wraps a Claude Managed Agents session and tracks observability data.

    Usage (sync stream)::

        from agentmetrics_anthropic import AgentMetricsSessionTracker

        tracker = AgentMetricsSessionTracker(api_key="am_...", agent_id="my-agent")

        with tracker.stream(client, session_id="sess_...") as stream:
            for event in stream:
                ...  # handle events as normal

    Usage (async stream)::

        async with tracker.astream(client, session_id="sess_...") as stream:
            async for event in stream:
                ...
    """

    def __init__(
        self,
        api_key: str,
        agent_id: str = "anthropic-agent",
        base_url: str = "http://localhost:8099",
    ) -> None:
        self._client   = HttpClient(api_key=api_key, base_url=base_url)
        self._agent_id = agent_id

    def _fresh_state(self, session_id: str) -> _SessionState:
        """INT-14: Create a fresh per-run state, resetting all counters and the _emitted flag."""
        return _SessionState(self._agent_id, session_id)

    def stream(self, client: Any, session_id: str, **kwargs: Any) -> _SyncStreamContext:
        return _SyncStreamContext(self._client, self._agent_id, client, session_id, kwargs, self._fresh_state)

    def astream(self, client: Any, session_id: str, **kwargs: Any) -> _AsyncStreamContext:
        return _AsyncStreamContext(self._client, self._agent_id, client, session_id, kwargs, self._fresh_state)

    def flush(self, timeout: float = 10.0) -> None:
        self._client.flush(timeout=timeout)


class _SyncStreamContext:
    def __init__(
        self,
        http: HttpClient,
        agent_id: str,
        client: Any,
        session_id: str,
        kwargs: dict,
        fresh_state: Any = None,
    ) -> None:
        self._http        = http
        self._agent_id    = agent_id
        self._client      = client
        self._session_id  = session_id
        self._kwargs      = kwargs
        self._fresh_state = fresh_state
        self._state: _SessionState | None = None

    def __enter__(self) -> _SyncTrackingIter:
        # INT-14: use the factory to create a fresh state, resetting all per-run fields
        if self._fresh_state is not None:
            self._state = self._fresh_state(self._session_id)
        else:
            self._state = _SessionState(self._agent_id, self._session_id)
        raw = self._client.beta.sessions.events.stream(
            self._session_id, **self._kwargs
        )
        return _SyncTrackingIter(raw, self._state, self._http)

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        if self._state:
            if exc_type is not None:
                self._state.status = "failed"
                self._state.error  = str(exc_val)[:500]
            if not self._state._emitted:
                self._state._emitted = True
                _emit(self._http, self._state)


class _SyncTrackingIter:
    def __init__(self, raw: Any, state: _SessionState, http: HttpClient) -> None:
        self._raw   = raw
        self._state = state
        self._http  = http
        self._done  = False

    def __iter__(self) -> _SyncTrackingIter:
        return self

    def __next__(self) -> Any:
        if self._done:
            raise StopIteration
        event = next(self._raw)
        done  = _process_event(self._state, event)
        if done:
            self._done = True
            if not self._state._emitted:
                self._state._emitted = True
                _emit(self._http, self._state)
        return event


class _AsyncStreamContext:
    def __init__(
        self,
        http: HttpClient,
        agent_id: str,
        client: Any,
        session_id: str,
        kwargs: dict,
        fresh_state: Any = None,
    ) -> None:
        self._http        = http
        self._agent_id    = agent_id
        self._client      = client
        self._session_id  = session_id
        self._kwargs      = kwargs
        self._fresh_state = fresh_state
        self._state: _SessionState | None = None

    async def __aenter__(self) -> _AsyncTrackingIter:
        # INT-14: use the factory to create a fresh state, resetting all per-run fields
        if self._fresh_state is not None:
            self._state = self._fresh_state(self._session_id)
        else:
            self._state = _SessionState(self._agent_id, self._session_id)
        raw = self._client.beta.sessions.events.stream(
            self._session_id, **self._kwargs
        )
        return _AsyncTrackingIter(raw, self._state, self._http)

    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        if self._state:
            if exc_type is not None:
                self._state.status = "failed"
                self._state.error  = str(exc_val)[:500]
            if not self._state._emitted:
                self._state._emitted = True
                _emit(self._http, self._state)


class _AsyncTrackingIter:
    def __init__(self, raw: Any, state: _SessionState, http: HttpClient) -> None:
        self._raw   = raw
        self._state = state
        self._http  = http
        self._done  = False

    def __aiter__(self) -> _AsyncTrackingIter:
        return self

    async def __anext__(self) -> Any:
        if self._done:
            raise StopAsyncIteration
        event = await self._raw.__anext__()
        done  = _process_event(self._state, event)
        if done:
            self._done = True
            if not self._state._emitted:
                self._state._emitted = True
                _emit(self._http, self._state)
        return event


def _emit(http: HttpClient, state: _SessionState) -> None:
    duration_ms = (time.monotonic() - state.start_ms) * 1000
    ev = AgentEndEvent(agent_id=state.agent_id, platform="anthropic")
    ev.session_id         = state.session_id
    ev.trace_id           = state.session_id
    ev.input_tokens       = state.input_tokens
    ev.output_tokens      = state.output_tokens
    ev.cache_read_tokens  = state.cache_read_tokens
    ev.cache_write_tokens = state.cache_write_tokens
    ev.llm_calls          = state.llm_calls
    ev.tool_calls         = state.tool_calls
    ev.tool_errors        = state.tool_errors
    ev.tool_names         = list(state.tool_names)
    ev.status             = state.status
    ev.duration_ms        = round(duration_ms, 2)
    ev.error              = state.error
    ev.model              = state.model
    ev.estimated_cost_usd = estimate_cost(
        state.model or "", state.input_tokens, state.output_tokens,
        state.cache_read_tokens, state.cache_write_tokens,
    ) or None
    http.fire_and_forget(ev.to_payload())
